File: backend/app/schemas/participation.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

ContactType = Literal["in_person", "phone", "video", "other"]


class ParticipationWrite(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    title: str = Field(min_length=1, max_length=180)
    occurred_on: date
    contact_type: ContactType = "in_person"
    conversation_partners: str = Field(default="", max_length=2000)
    location: str = Field(default="", max_length=240)
    duration_minutes: int | None = Field(default=None, gt=0, le=1440)
    report: str = Field(min_length=1, max_length=30000)
    agreements: str = Field(default="", max_length=10000)
    participant_ids: list[str] = Field(min_length=1, max_length=100)
    project_ids: list[str] = Field(min_length=1, max_length=100)

    @field_validator("participant_ids", "project_ids")
    @classmethod
    def unique_ids(cls, values: list[str]) -> list[str]:
        if any(not value.strip() or len(value) > 36 for value in values):
            raise ValueError("Selecteer geldige personen en projecten.")
        if len(set(values)) != len(values):
            raise ValueError("Selecteer iedere persoon en ieder project maximaal één keer.")
        return values


class ParticipationUpdate(ParticipationWrite):
    expected_row_version: int = Field(ge=1)


class ParticipationArchive(BaseModel):
    model_config = ConfigDict(extra="forbid")
    archived: bool
    expected_row_version: int = Field(ge=1)


class ParticipationOption(BaseModel):
    id: str
    name: str
    selectable: bool = True


class ParticipationMeta(BaseModel):
    participants: list[ParticipationOption]
    projects: list[ParticipationOption]
    current_user_id: str


class ParticipationResponse(BaseModel):
    id: str
    title: str
    occurred_on: date
    contact_type: ContactType
    conversation_partners: str
    location: str
    duration_minutes: int | None
    report: str
    agreements: str
    participants: list[ParticipationOption]
    projects: list[ParticipationOption]
    created_by_name: str
    created_at: datetime
    updated_at: datetime
    archived_at: datetime | None
    row_version: int
    can_edit: bool


class ParticipationFilters(BaseModel):
    model_config = ConfigDict(extra="forbid")
    query: str = Field(default="", max_length=200)
    date_from: date | None = None
    date_to: date | None = None
    project_id: str | None = None
    participant_id: str | None = None
    archived: bool = False
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=25, ge=1, le=100)
    format: Literal["csv", "markdown", "json"] = "csv"

    @model_validator(mode="after")
    def date_range(self):
        if self.date_from and self.date_to and self.date_from > self.date_to:
            raise ValueError("De begindatum moet vóór of op de einddatum liggen.")
        return self


class ParticipationList(BaseModel):
    items: list[ParticipationResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/lib/api/participation.ts
import { request, requestBlob } from "./client";

export type ParticipationOption = { id: string; name: string; selectable: boolean };
export type ContactType = "in_person" | "phone" | "video" | "other";
export type ParticipationDraft = {
  title: string; occurred_on: string; contact_type: ContactType; conversation_partners: string;
  location: string; duration_minutes: number | null; report: string; agreements: string;
  participant_ids: string[]; project_ids: string[];
};
export type ParticipationMoment = Omit<ParticipationDraft, "participant_ids" | "project_ids"> & {
  id: string; participants: ParticipationOption[]; projects: ParticipationOption[];
  created_by_name: string; created_at: string; updated_at: string; archived_at: string | null;
  row_version: number; can_edit: boolean;
};
export type ParticipationMeta = { participants: ParticipationOption[]; projects: ParticipationOption[]; current_user_id: string };
export type ParticipationFilters = {
  query: string; date_from: string; date_to: string; project_id: string; participant_id: string; archived: boolean;
};
export type ParticipationList = { items: ParticipationMoment[]; total: number; page: number; page_size: number };
export type ParticipationExportFormat = "csv" | "markdown" | "json";

function queryString(values: Record<string, string | number | boolean>): string {
  return new URLSearchParams(Object.entries(values).filter(([, value]) => value !== "").map(([key, value]) => [key, String(value)])).toString();
}
const base = "/participatiemomenten";
export const getParticipationMeta = () => request<ParticipationMeta>(`${base}/meta`);
export const listParticipationMoments = (filters: ParticipationFilters, page = 1) =>
  request<ParticipationList>(`${base}?${queryString({ ...filters, page, page_size: 25 })}`);
export const getParticipationMoment = (id: string) => request<ParticipationMoment>(`${base}/${encodeURIComponent(id)}`);
export const createParticipationMoment = (draft: ParticipationDraft) =>
  request<ParticipationMoment>(base, { method: "POST", body: JSON.stringify(draft) });
export const updateParticipationMoment = (id: string, draft: ParticipationDraft, version: number) =>
  request<ParticipationMoment>(`${base}/${encodeURIComponent(id)}`, { method: "PUT", body: JSON.stringify({ ...draft, expected_row_version: version }) });
export const archiveParticipationMoment = (id: string, archived: boolean, version: number) =>
  request<ParticipationMoment>(`${base}/${encodeURIComponent(id)}/archive`, { method: "PATCH", body: JSON.stringify({ archived, expected_row_version: version }) });
export const exportParticipationMoments = (filters: ParticipationFilters, format: ParticipationExportFormat) =>
  requestBlob(`${base}/export?${queryString({ ...filters, format })}`);

export function participationError(error: unknown): string {
  if (!(error instanceof Error)) return "Er ging iets mis. Probeer opnieuw.";
  try {
    const { detail } = JSON.parse(error.message);
    if (typeof detail === "string") return detail;
    if (Array.isArray(detail)) return detail.map(item => item.msg).join(" ");
  } catch { /* Network and plain error messages are handled below. */ }
  return "Het verzoek kon niet worden afgerond. Controleer je verbinding en probeer opnieuw.";
}
